File: tictactoe.java
import java.util.Scanner;

public class TicTacToe {
    public static char[][] initBoard() {
        char[][] board = new char[3][3];
        char cell = '1';
        for (int i = 0; i < 3; i++) {
            for (int j = 0; j < 3; j++) {
                board[i][j] = cell++;
            }
        }
        return board;
    }

    public static void printBoard(char[][] board) {
        System.out.println();
        for (int i = 0; i < 3; i++) {
            System.out.printf(" %c | %c | %c ", board[i][0], board[i][1], board[i][2]);
            System.out.println();
            if (i != 2) {
                System.out.println("---+---+---");
            }
        }
        System.out.println();
    }

    public static boolean checkWin(char[][] board) {
        // rows and columns
        for (int i = 0; i < 3; i++) {
            if (board[i][0] == board[i][1] && board[i][1] == board[i][2]) {
                return true;
            }
            if (board[0][i] == board[1][i] && board[1][i] == board[2][i]) {
                return true;
            }
        }
        // diagonals
        if (board[0][0] == board[1][1] && board[1][1] == board[2][2]) {
            return true;
        }
        if (board[0][2] == board[1][1] && board[1][1] == board[2][0]) {
            return true;
        }
        return false;
    }

    public static boolean isDraw(char[][] board) {
        for (int i = 0; i < 3; i++) {
            for (int j = 0; j < 3; j++) {
                if (board[i][j] >= '1' && board[i][j] <= '9') {
                    return false;
                }
            }
        }
        return true;
    }

    public static boolean makeMove(char[][] board, char player, int choice) {
        char target = (char) (choice + '0');
        for (int i = 0; i < 3; i++) {
            for (int j = 0; j < 3; j++) {
                if (board[i][j] == target) {
                    board[i][j] = player;
                    return true;
                }
            }
        }
        return false;
    }

    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);
        char player = 'X';
        char[][] board = initBoard();
        System.out.println("Tic Tac Toe (Player X vs Player O)");
        printBoard(board);

        while (true) {
            System.out.printf("Player %c, enter a cell number (1-9): ", player);
            int choice;
            try {
                choice = Integer.parseInt(scanner.nextLine());
            } catch (NumberFormatException e) {
                System.out.println("Invalid input. Exiting.");
                break;
            }

            if (choice < 1 || choice > 9) {
                System.out.println("Invalid input. Exiting.");
                break;
            }

            if (!makeMove(board, player, choice)) {
                System.out.println("Cell already taken. Try again.");
                continue;
            }

            printBoard(board);

            if (checkWin(board)) {
                System.out.printf("Player %c wins!%n", player);
                break;
            } else if (isDraw(board)) {
                System.out.println("It's a draw.");
                break;
            } else {
                player = (player == 'X') ? 'O' : 'X';
            }
        }
        scanner.close();
    }
}

File: tictactoe.kt
import java.util.Scanner

fun initBoard(): Array<CharArray> {
    val board = Array(3) { CharArray(3) }
    var cell = '1'
    for (i in 0..2) {
        for (j in 0..2) {
            board[i][j] = cell
            cell++
        }
    }
    return board
}

fun printBoard(board: Array<CharArray>) {
    println()
    for (i in 0..2) {
        println(" ${board[i][0]} | ${board[i][1]} | ${board[i][2]} ")
        if (i != 2) {
            println("---+---+---")
        }
    }
    println()
}

fun checkWin(board: Array<CharArray>): Boolean {
    for (i in 0..2) {
        if (board[i][0] == board[i][1] && board[i][1] == board[i][2]) {
            return true
        }
        if (board[0][i] == board[1][i] && board[1][i] == board[2][i]) {
            return true
        }
    }
    if (board[0][0] == board[1][1] && board[1][1] == board[2][2]) {
        return true
    }
    if (board[0][2] == board[1][1] && board[1][1] == board[2][0]) {
        return true
    }
    return false
}

fun isDraw(board: Array<CharArray>): Boolean {
    for (i in 0..2) {
        for (j in 0..2) {
            if (board[i][j] in '1'..'9') {
                return false
            }
        }
    }
    return true
}

fun makeMove(board: Array<CharArray>, player: Char, choice: Int): Boolean {
    val target = choice.toString()[0]
    for (i in 0..2) {
        for (j in 0..2) {
            if (board[i][j] == target) {
                board[i][j] = player
                return true
            }
        }
    }
    return false
}

fun main() {
    val scanner = Scanner(System.`in`)
    var player = 'X'
    var gameOver = false
    val board = initBoard()
    println("Tic Tac Toe (Player X vs Player O)")
    printBoard(board)

    while (!gameOver) {
        print("Player $player, enter a cell number (1-9): ")
        val input = scanner.nextLine()
        val choice = input.toIntOrNull()
        if (choice == null || choice !in 1..9) {
            println("Invalid input. Exiting.")
            break
        }

        if (!makeMove(board, player, choice)) {
            println("Cell already taken. Try again.")
            continue
        }

        printBoard(board)

        when {
            checkWin(board) -> {
                println("Player $player wins!")
                gameOver = true
            }
            isDraw(board) -> {
                println("It's a draw.")
                gameOver = true
            }
            else -> {
                player = if (player == 'X') 'O' else 'X'
            }
        }
    }
}
